presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java

```diff
@@ -221,38 +228,86 @@ public static HiveSplitSource bucketed(
-                            // Otherwise, `queue.finish` may not be invoked if `finished` is set while the lambda above is being evaluated.
+                        return buckets.computeIfAbsent(bucketNumber.getAsInt(), ignored -> new BucketedSplits(estimatedOutstandingSplitsPerBucket, executor));
+                    }
+
+                    // By adding this synchronization wrapper, we could reduce synchronization granularity to per-bucket level.
+                    class BucketedSplits
+                    {
+                        private final List<InternalHiveSplit> discoveredHiveSplits;
+                        private final AsyncQueue<InternalHiveSplit> queue;
+                        private Optional<ListenableFuture<?>> lastBorrowBatchAsyncFuture = Optional.empty();
+
+                        public BucketedSplits(int estimatedQueueSize, Executor executor)
+                        {
+                            this.discoveredHiveSplits = new ArrayList<>();
+                            this.queue = new AsyncQueue<>(estimatedQueueSize, requireNonNull(executor, "executor is null"));
+                        }
+
+                        public synchronized void offer(InternalHiveSplit internalHiveSplit)
+                        {
+                            discoveredHiveSplits.add(internalHiveSplit);
+                            queue.offer(internalHiveSplit);
+                        }
+
+                        public synchronized <O> ListenableFuture<O> borrowBatchAsync(int maxSize, Function<List<InternalHiveSplit>, BorrowResult<InternalHiveSplit, O>> function)
+                        {
+                            checkState(!lastBorrowBatchAsyncFuture.isPresent() || lastBorrowBatchAsyncFuture.get().isDone(), "last borrowBatchAsync must finish if exists");
+                            ListenableFuture<O> borrowBatchAsyncFuture = queue.borrowBatchAsync(maxSize, function);
+                            lastBorrowBatchAsyncFuture = Optional.of(borrowBatchAsyncFuture);
+                            return borrowBatchAsyncFuture;
+                        }
+
+                        public synchronized void finish()
+                        {
                             queue.finish();
                         }
-                        return queue;
+
+                        public synchronized boolean isFinished()
+                        {
+                            return queue.isFinished();
+                        }
+
+                        public synchronized int rewind()
+                        {
+                            if (lastBorrowBatchAsyncFuture.isPresent()) {
+                                lastBorrowBatchAsyncFuture.get().cancel(true);
+                                lastBorrowBatchAsyncFuture = Optional.empty();
+                            }
+                            int revivedSplitCount = discoveredHiveSplits.size() - queue.size();
+                            queue.clear();
```

Comment:
`AsyncQueue.borrowCount` is tricky, since in theory it can be modified in Future (although already cancelled). 

What about create a new `AsynQueue` instead of clearing the old `AsyncQueue` ? :) 

Thoughts? @arhimondr , @rschlussel

Reply:
Right now do we ever cancel something that we actually care that it doesn't finish (for correctness reasons)? I didn't see anything, but if there is I'd be curious what its behavior is. Since there are no guarantees on what happens once the future is cancelled, I agree that it would be safer to create a new AsyncQueuue